azure-mgmt-resources/src/main/java/com/azure/management/resources/fluentcore/utils/ETagState.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+
+package com.azure.management.resources.fluentcore.utils;
+
+public class ETagState {
+    private boolean doImplicitETagCheckOnCreate;
+    private boolean doImplicitETagCheckOnUpdate;
+    private String eTagOnCreate;
+    private String eTagOnUpdate;
+    private String eTagOnDelete;
+
+    public ETagState withImplicitETagCheckOnCreate() {
+        this.doImplicitETagCheckOnCreate = true;
+        return this;
+    }
+
+    public ETagState withImplicitETagCheckOnUpdate() {
```

Comment:
This action is due to how we update/delete the resource. For update action, we operate on resource itself like `resource.update().apply()`. But for delete, we operate on resources/parent level like `azure.resources.delete(resourceId)`.

Reply:
You just have a `deleteResourceAsync` in DnsRecordSetImpl.java line 344, isn't it (I assume this is the reason you are adding `eTagOnDelete`)? Is it to be a common approach, or just one time case?

Reply:
It's common. ETag value for `Delete` will always be explicit.

Reply:
So the question, why it need to be explicit? Why it cannot be implicit? (of course I know existing code do `delete(id, etag)`, but the question is should we, any better approach if we already had the resource object?)

Another question is why add `eTagOnDelete` while the common approach is `delete(id, etag)`?

Reply:
It's about the trade-off for the switch. I didn't find out good solution to make it implicit though we are able to do that.

For `etagOnDelete`, it's because record set is sub resource. The working flow is that we label the record set `ToBeRemoved` and send delete request later. Hence we have to store the etag for Delete somewhere.

Reply:
This is a *DESIGN* task. It is all about to list the trade-off explicitly, and then the reason to choose the approach.

E.g. what if `delete(FooResource, ETag.IfNotChanged)`? What if `FooResource.withEtagCheck().delete()`?

For `etagOnDelete`, the question is whether this is common usage. Here this is a shared class. If the usage is only for record set then it should only in dns. Again your investigation is needed.

Reply:
Sure, I just list the points I considered for this plan as below:
- existing `withETagCheck()` and `withETagCheck(etagValue)`
In current implementation, the default option is without appending `withETagCheck()`/`withETagCheck(etagValue)`. When appending `withETagCheck()`, it's implicit etag. For `withETagCheck(etagValue)`, it allows to pass explicit etag.

It would make sense to keep `withETagCheck(etagValue)` as the users might want to enable weak validation for etag in request header. 
- Based on above, the design is supposed to keep both explicit and implicit etag so that users could pick their options whether to send conditional request and which way they prefer to pass etag value. For my understanding, this should also apply in Delete operation as well as that in Update operation. The current method `delete(id, etag)` already meets the requirement.
- For `etagOnDelete`, it's common usage for only sub resource. For this, maybe we could separate the `ETagState` into `GeneralETagState` and `SubResourceETagState`. `GeneralETagState` will be the base class with only create and update related. `SubResourceETagState` will be sub class extends with `etagOnDelete`. So `SubResourceETagState` could be available for other sub resources if applicable in the future.

Reply:
based on the design guideline, user is default opt-in for etag check. and default unconditional. With that a implicit etag delete would make sense, same as create/update. I think this point we're aligned.  for explicit etag check, problem is what's best interface. Concern from @xccc-msft is for subresource to pass the etag value. let's talk this offline.